### ISSUE TITLE
(fchollet): it should be possible (and highly desirable) to support 

### ISSUE THREAD

**Issue by piascikj** [imdone/tensorflow#3242]
`add_loss` in eager mode. This allows great convenience and flexibility
in defining custom losses on the fly (e.g. in VAEs).
Simply appending the loss value to `self._losses`
is the correct behavior.
The only caveat is that we need to force the user to only call
`add_loss` from inside a model or Layer's `call` method
(otherwise the loss computation cannot be backproped through).

----
*opened via [imdone.io](https://imdone.io) from a code comment on [693b339ab2f062ec5bbb29f976c5d1fd94fbffa5](https://github.com/imdone/tensorflow/commit/693b339ab2f062ec5bbb29f976c5d1fd94fbffa5) by Francois Chollet fchollet@google.com*

----

https://github.com/imdone/tensorflow/blob/8cb6e535f6ea14b380aa86d425169adad682cb8c/tensorflow/python/keras/_impl/keras/engine/base_layer.py#L382-L394
